src/world.c

```diff
@@ -10,22 +11,131 @@ void create_world(int p, int q, world_func func, void *arg) {
             int h = f * mh;
             int w = 1;
             int t = 12;
+			int tc = 25;
+			if (b >= 0.66) {
+				w = 74;
+			}
             if (h <= t) {
                 h = t;
                 w = 2;
             }
-            // sand and grass terrain
+			if (h > tc) {
+				w = 9;
+			}
+            // terrain 
             for (int y = 0; y < h; y++) {
-                func(x, y, z, w * flag, arg);
+				int u = w;
+				if (y < h - 1){
+					if (simplex3(
+                        x * 0.01, y * 0.1, z * 0.01, 8, 0.5, 2) > 0.8) {
+						u = 78 + simplex2(x * 0.1, z * 0.1, 4, 0.8, 2) * 5;
+					}
+					switch(u) {
+						case 1:
+						case 9:
+							u = 7;
+							break;
+						case 74:
+							u = 83;
+							break;
+					}
+				}
+               	func(x, y, z, u * flag, arg);
             }
+			//-------------------
+			//---Requirement 4---
+			if (w == 2) {
+				if (SHOW_PLANTS) {
+					// desert plants
+					if (simplex2(-x * 0.1, z * 0.1, 3, 0.8, 2) > 0.8) {
+						int w = 64 + simplex2(x * 0.1, z * 0.1, 4, 0.2, 2) * 3;
+						func(x, h, z, w * flag, arg);
+					}
+				}
+				// desert 'trees'
+				int ok = SHOW_TREES;
+				if (ok && simplex2(x, z, 6, 0.5, 2) > 0.87) {
+                    for (int y = h; y < h + 3; y++) {
+                        func(x, y, z, 75, arg);
+                    }
+                }
+			}
+			if (w == 9) {
+				if (SHOW_PLANTS) {
+					// tundra plants
+					if (simplex2(-x * 0.1, z * 0.1, 3, 0.8, 2) > 0.7) {
+						int w = 67 + simplex2(x * 0.1, z * 0.1, 4, 0.3, 2) * 3;
+						func(x, h, z, w * flag, arg);
+					}
+				}
+				// tundra trees
+				int ok = SHOW_TREES;
+				if (dx - 4 < 0 || dz - 4 < 0 ||
+                    dx + 4 >= CHUNK_SIZE || dz + 4 >= CHUNK_SIZE)
+                {
+                    ok = 0;
+                }
+				if (ok && simplex2(x, z, 6, 0.5, 2) > 0.84) {
+                    for (int y = h + 2; y < h + 8; y++) {
+                        for (int ox = -4; ox <= 4; ox++) {
+                            for (int oz = -4; oz <= 4; oz++) {
+								int d = ((ABS(ox) + ABS(oz))) + (y - (h + 2));
+								if (d <= 5) {
+                            		func(x + ox, y, z + oz, 85, arg);
+								}
+                            }
+                        }
+                    }
+                    for (int y = h; y < h + 7; y++) {
+                        func(x, y, z, 76, arg);
+                    }
+                }
+			}
+			if (w == 74){
+				if (SHOW_PLANTS) {
+					// jungle plants
+					if (simplex2(-x * 0.1, z * 0.1, 3, 0.8, 2) > .7) {
+						int w = 70 + simplex2(x * 0.1, z * 0.1, 4, 0.1, 2) * 3;
+						func(x, h, z, w * flag, arg);
+					}
+				}
+				// jungle trees
+				int ok = SHOW_TREES;
+				if (dx - 4 < 0 || dz - 4 < 0 ||
+                    dx + 4 >= CHUNK_SIZE || dz + 4 >= CHUNK_SIZE)
+                {
+                    ok = 0;
+                }
+				if (ok && simplex2(x, z, 6, 0.5, 2) > 0.8) {
+                    for (int y = h + 3; y < h + 8; y++) {
```

Comment:
Whenever I see repetitive code such as the 3 nested for loops here and on line 92, I typically try to find a way to create a function instead of retyping it over and over. Since you are doing different algebra inside each I don't know for sure if there is an easy way to do that unless you make more functions containing the algebra for various trees, and call each from inside the loop function as well, but still maybe inpractical. If more trees were to be added in the future it might make things easier moving forward.

Reply:
That's a good point, I'll go back and make separate methods for deciduous/coniferous trees that can be passed the index for the block to use.